docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,350 @@
+
+**Modify Page Attributes**
+
+The EMM clients may call sgx_mm_modify/sgx_mm_modify_ex to request permissions and/or page type changes for pages in existing allocations.
+
+
+## Notes on Internal Design ##
+
+The enclave memory manager keeps track of memory allocation and layout info inside enclave address range (ELRANGE) using an internal structure called the Enclave Memory Area (EMA) List. The EMA and EMA list are considered private data structure of memory manager, and their internals are not exposed in client-facing APIs.
+- The EMA list tracks all memory regions in use (reserved, committed, commit-on-demand) in ELRANGE.
+- Ranges in ELRANGE not tracked by an EMA are considered free and ready for new allocations
+- The EMM labels certain EMAs reserved for runtime or its internal usage and make them not accessible from public APIs.
+- A thread calling an EMM API on an EMA with operation pending in another thread will wait until the pending operation is finished. 
+
+**Assumptions:**
+
+- When enclave is loaded, the OS has reserved the whole address range covered by ELRANGE. It is assumed the host app will not remap any part of this reserved range.
+  - In the future, when the new feature that allows loading an enclave with base address at zero, we may support partial ELRANGE reserved by the OS.  
+- Memory manager does not check EPC pressure, or proactively trim pages when EPC runs low. The OS can reclaim EPC pages when EPC running low or cgroups threshold reached
+- Memory manager does not maintain and recycle committed then freed pages
+  - Whenever a page is freed (via dealloc or uncommit API), it is trimmed from enclave and need be re-allocated and committed before re-use.
+  - Owner of a region can re-purpose a sub-region of it by calling sgx_mm_modify_type/permissions to split out the sub-region to be reused.
+- Memory manager does not call back client for #GP handling. Memory manager code will ensure itself would not cause #GP, and only register a #PF handler with the enclave global exception handler registry. Clients wish to handle #GP can register its own exception handler with the global handler registry.
+
+
+Public APIs
+-----------------
+
+### sgx_mm_alloc
+
+Allocate a new memory region inside enclave and optionally register a custom page fault handler for the region
+
+```
+/**
+ * Page fault (#PF) info reported in the SGX SSA MISC region.
+ */
+typedef struct _sgx_pfinfo
+{
+    uint64_t maddr;                     //address for #PF
+    union _pfec {                          
+        uint32_t errcd;      
+        struct {//PFEC bits
+            uint32_t p       : 1;    // P flag
+            uint32_t rw      : 1;    // RW access flag, 0 for read, 1 for write
+            uint32_t         : 13;   // U/S, I/O, PK and reserved bits not relevant for SGX PF
+            uint32_t sgx     : 1;    // SGX bit
+            uint32_t         : 16;   // reserved bits
+        };
+    } pfec;
+    uint32_t reserved; 
+} sgx_pfinfo;
```

Comment:
Note from OE: the similar data structure will be supported by OE as part of #3984.